### PR TITLE
ISSUE-1108 Implement OpenAI Mailet (version 1)

### PR DESCRIPTION
**Pull request by quantranhong1999**
To create a mail bot calling OpenAI APIs...

No chat memory management yet. It means GPT won't understand the context of the previous chat.

[open ai mailet version 1.webm](https://github.com/user-attachments/assets/0452520d-47e3-444d-a0af-b269b39aad1a)

### CODE REVIEW

**Review by Arsnael — COMMENTED** [review]
Great I know how to cook an egg now :D 

Nothing to say about the code. Maybe antora doc missing though? But looks great otherwise!

File: demo/tmail/mailetcontainer.xml [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+            <mailet match="All" class="VacationMailet"/>
+            <mailet match="All" class="Sieve"/>
+            <mailet match="All" class="AddDeliveredToHeader"/>
+            <mailet match="All" class="org.apache.james.jmap.mailet.filter.JMAPFiltering"/>
+            <mailet match="All" class="com.linagora.tmail.mailets.TmailLocalDelivery">
+                <consume>false</consume>
+            </mailet>
+
+            <!-- Put the OpenAIMailet after LocalDelivery so the GPT reply would come after the asking question -->
+            <mailet match="com.linagora.tmail.mailet.RecipientsContain=gpt@tmail.com" class="com.linagora.tmail.mailet.OpenAIMailet">
```
Comment (chibenwa) [inline demo/tmail/mailetcontainer.xml:96]: How about `RecipientIs=gpt@tmail.com` ?

File: tmail-backend/mailets/src/main/java/com/linagora/tmail/mailet/OpenAIMailet.java [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+        MimeMessage reply = (MimeMessage) mail.getMessage()
+            .reply(true);
+
+        stripGptFromRecipients(reply, Message.RecipientType.TO);
+        stripGptFromRecipients(reply, Message.RecipientType.CC);
+        stripGptFromRecipients(reply, Message.RecipientType.BCC);
+
+        reply.setSentDate(new Date());
+        reply.setFrom(gptAddress.asString());
+        reply.setText(answer);
```
Comment (chibenwa) [inline tmail-backend/mailets/src/main/java/com/linagora/tmail/mailet/OpenAIMailet.java:101]: Ideally we should quote the previous answer if not too complex

```
Boil it for 10 minutes

On 15th july grandma <gm@toto.com> wrote:
> How to cook an egg?
```

If too complex I would agree with what had been done today.

File: tmail-backend/mailets/src/main/java/com/linagora/tmail/mailet/OpenAIMailet.java [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+    private MailAddress gptAddress;
+
+    @Override
+    public void service(Mail mail) throws MessagingException {
+        if (gptLoopDetected(mail)) {
+            return;
+        }
+
+        try {
+            String prompt = MimeMessageUtil.asString(mail.getMessage());
```
Comment (chibenwa) [inline tmail-backend/mailets/src/main/java/com/linagora/tmail/mailet/OpenAIMailet.java:72]: Do we still and sanitize HTML?
Is mime multipart (mixed, related, alternative) handled?

File: tmail-backend/mailets/src/main/java/com/linagora/tmail/mailet/OpenAIMailet.java [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+public class OpenAIMailet extends GenericMailet {
+    public static final String API_KEY_PARAMETER_NAME = "apiKey";
+    public static final String GPT_ADDRESS_PARAMETER_NAME = "gptAddress";
+
+    private ChatLanguageModel openAiModel;
+    private MailAddress gptAddress;
+
+    @Override
+    public void service(Mail mail) throws MessagingException {
+        if (gptLoopDetected(mail)) {
```
Comment (chibenwa) [inline tmail-backend/mailets/src/main/java/com/linagora/tmail/mailet/OpenAIMailet.java:67]: ^^

File: demo/tmail/mailetcontainer.xml (hunk 1 above)
Comment (quantranhong1999) [reply demo/tmail/mailetcontainer.xml:96]: That matcher would remove other recipients -> they can not receive the answer.

Comment (chibenwa) [reply demo/tmail/mailetcontainer.xml:96]: Yes I got it when reading the matcher code.

File: tmail-backend/mailets/src/main/java/com/linagora/tmail/mailet/OpenAIMailet.java (hunk 3 above)
Comment (quantranhong1999) [reply tmail-backend/mailets/src/main/java/com/linagora/tmail/mailet/OpenAIMailet.java:72]: This method writes the `MimeMessage` to a outputstream.

Result:
```
Received: from localhost (HELO james.org) ([127.0.0.1])
          by hp-quanth (JAMES SMTP Server ) with SMTP ID 9f40f9d1
          for <gpt@james.org>;
          Fri, 19 Jul 2024 09:12:44 +0700 (ICT)
Date: Fri, 19 Jul 2024 09:12:44 +0700 (ICT)
Sender: bob@james.org
To: gpt@james.org
Message-ID: <654233182.0.1721355164402@hp-quanth>
Subject: How can I cook an egg?
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 7bit

I do not know how to cook an egg. Please help me.
```

It seems to be an eml to me. And OpenAI seems smart enough to understand the important part.

Before, I tried to extract only the subject and body, but it seems complicated to extract the body and I am not confident handles well all the cases.
image

So a bit of no brain here...

Comment (chibenwa) [reply tmail-backend/mailets/src/main/java/com/linagora/tmail/mailet/OpenAIMailet.java:72]: Discard the subject

How about using `MessageContentExtractor` ?

**Review by chibenwa — APPROVED** [review]

File: tmail-backend/tmail-third-party/open-ai/README.md [hunk 5]
```diff
@@ -0,0 +1,22 @@
+# OpenAI third party
```
Comment (Arsnael) [inline tmail-backend/tmail-third-party/open-ai/README.md:1]: Said it before but... little antora doc with this? Then I'm ok :)

**Review by vttranlina — APPROVED** [review]
read it

**Review by Arsnael — APPROVED** [review]

**Comment by quantranhong1999** [comment]
squashed

**Review by chibenwa — APPROVED** [review]

**Pull request merged**